File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.kt
import java.util.*
import kotlin.collections.HashSet
import kotlin.math.abs

class Solution {
    lateinit var arr: Array<MutableList<Int>>

    fun solution(n: Int, wires: Array<IntArray>): Int {
        var ans = n

        arr = Array(n) { mutableListOf() }
        wires.forEach {
            arr[it[0] - 1].add(it[1] - 1)
            arr[it[1] - 1].add(it[0] - 1)
        }
        wires.forEach { ans = minOf(ans, abs(2 * bfs(it[0] - 1, it[1] - 1) - n)) }

        return ans
    }

    fun bfs(start: Int, end: Int): Int {
        val list = arrayListOf(start)
        val que: Queue<Int> = LinkedList()

        arr[start].filter { it != end }.let {
            list.addAll(it)
            que.addAll(it)
        }

        while (que.isNotEmpty()) {
            arr[que.poll()].filter { list.contains(it).not() }.let {
                list.addAll(it)
                que.addAll(it)
            }
        }

        return list.size
    }
}

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    global graph
    
    graph = [[] for _ in range(n + 1)]
    ans = n
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
        
    for a, b in wires:
        ans = min(ans, abs(2 * bfs(a, b) - n))
    
    return ans

def bfs(start, end):
    list = [start]
    q = deque([])
    
    for i in graph[start]:
        if i != end:
            list.append(i)
            q.append(i)
        
    while q:
        for i in graph[q.popleft()]:
            if i not in list:
                list.append(i)
                q.append(i)
                
    return len(list)